File: objc/kwestkingdom/src/Shield.m
/**
 * Copyright 2009 David Couzelis
 * 
 * This file is part of "Kwest Kingdom".
 * 
 * "Kwest Kingdom" is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * "Kwest Kingdom" is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with "Kwest Kingdom".  If not, see <http://www.gnu.org/licenses/>.
 */
#import "Shield.h"


typedef enum {
  SHIELD_HOLD_STATE,
  SHIELD_AWAY_STATE
} SHIELD_STATE;


@implementation Shield


- init {
  
  self = [super init];
  
  if (self) {
    
    holdAnimation = [[Animation alloc] init];
    [holdAnimation addFrame: getImage(IMG_SHIELD_1)];
    [holdAnimation addFrame: getImage(IMG_SHIELD_2)];
    [holdAnimation addFrame: getImage(IMG_SHIELD_3)];
    [holdAnimation addFrame: getImage(IMG_SHIELD_4)];
    [holdAnimation setLoop: YES];
    [holdAnimation setSpeed: 6];
	
    [self toAwayState];
    
  }
  
  return self;
  
}


- (void) dealloc {
  [holdAnimation release];
  [super dealloc];
}


- toHoldState {
  state = SHIELD_HOLD_STATE;
  animation = holdAnimation;
  return self;
}


- toAwayState {
  state = SHIELD_AWAY_STATE;
  animation = nil;
  return self;
}


- (BOOL) held {
  if (state == SHIELD_HOLD_STATE) {
    return YES;
  }
  return NO;
}


@end

File: allegro5/src/Shield.cpp
#include "resources.h"
#include "Shield.h"


typedef enum {
  SHIELD_HOLD_STATE,
  SHIELD_AWAY_STATE
} SHIELD_STATE;


Shield::Shield() {
  
  init_anim(&hold_anim, ON, 6);
  add_frame(&hold_anim, IMG("shield_1.bmp"));
  add_frame(&hold_anim, IMG("shield_2.bmp"));
  add_frame(&hold_anim, IMG("shield_3.bmp"));
  add_frame(&hold_anim, IMG("shield_4.bmp"));
  
  this->toAwayState();
}


Shield::~Shield() {
}


void Shield::toHoldState() {
  state = SHIELD_HOLD_STATE;
  anim = &hold_anim;
  reset_anim(anim);
}


void Shield::toAwayState() {
  state = SHIELD_AWAY_STATE;
  anim = NULL;
}


bool Shield::held() {
  if (state == SHIELD_HOLD_STATE) {
    return true;
  }
  return false;
}

